packages/notebook-extension/src/index.ts

```diff
@@ -336,6 +334,116 @@ export const commandEditItem: JupyterFrontEndPlugin<void> = {
   }
 };
 
+/**
+ * A plugin providing export commands in the main menu and command palette
+ */
+export const exportPlugin: JupyterFrontEndPlugin<void> = {
+  id: '@jupyterlab/notebook-extension:export',
+  autoStart: true,
+  requires: [ITranslator, INotebookTracker],
+  optional: [IMainMenu, ICommandPalette],
+  activate: (
+    app: JupyterFrontEnd,
+    translator: ITranslator,
+    tracker: INotebookTracker,
+    mainMenu: IMainMenu | null,
+    palette: ICommandPalette | null
+  ) => {
+    const trans = translator.load('jupyterlab');
+    const { commands, shell } = app;
+    const services = app.serviceManager;
+
+    const isEnabled = (): boolean => {
+      return Private.isEnabled(shell, tracker);
+    };
+
+    commands.addCommand(CommandIDs.exportToFormat, {
+      label: args => {
+        const formatLabel = args['label'] as string;
+        return args['isPalette']
+          ? trans.__('Export Notebook: %1', formatLabel)
+          : formatLabel;
+      },
+      execute: args => {
+        const current = getCurrent(tracker, shell, args);
+
+        if (!current) {
+          return;
+        }
+
+        const url = PageConfig.getNBConvertURL({
+          format: args['format'] as string,
+          download: true,
+          path: current.context.path
+        });
+        const child = window.open('', '_blank');
+        const { context } = current;
+
+        if (child) {
+          child.opener = null;
+        }
+        if (context.model.dirty && !context.model.readOnly) {
+          return context.save().then(() => {
+            child?.location.assign(url);
+          });
+        }
+
+        return new Promise<void>(resolve => {
+          child?.location.assign(url);
+          resolve(undefined);
+        });
+      },
+      isEnabled
+    });
+
+    if (mainMenu) {
```

Comment:
Should this check happen after getting the response from the server?

Otherwise if the menu is not provided, it looks like the command would not be added to the command palette either? (the `if (palette)` check below)

Reply:
In the original code, this section (lines 399-442) is all bundled into populateMenus, which is only called after checking for ```if (mainMenu)``` in its parent function. 

https://github.com/jupyterlab/jupyterlab/blob/c1f0b02748575a8fae0b4e3cede3a9060a94a0ea/packages/notebook-extension/src/index.ts#L951-L962

That's why I put the check here before anything gets added--I don't have context on the original code so I'm not sure if there's a specific reason it's all bundled together. Should it be different here after pulling it out?

Reply:
Ah right, thanks for looking it up :+1: 

Not sure there was a specific reason indeed to have it that way. It just caught attention since usually the menu and palettes are handled as two independent components.

But it would be totally fine to pull it out in a separate change too.

Reply:
@jtpio just moved it down! Let me know if you think that's a good spot for it, or if I should reorganize it differently.